### PR TITLE
add polygon hook and ism

### LINKED ISSUE #2847: Polygon Hooks and ISMs
This ticket encapsulates a hook and ISM that allows a Hyperlane message to be transported via Polygon's native PoS bridge. You can read more about it here https://wiki.polygon.technology/docs/category/bridges/

Submission of this bounty should include a demo on a testnet or mainnet with tx hashes.

### PR DESCRIPTION
**Pull request by NOOMA-42**
### Description

polygon hook and ism

### Related issues

- Fixes #2847 
- Documented with https://github.com/hyperlane-xyz/v3-docs/pull/31

### Backward compatibility

Yes

### Testing

`$ hyperlane-monorepo/solidity/: forge test -vvv --match-path test/isms/PolygonPosIsm.t.sol`

### Note
Per this [discord discussion](https://discord.com/channels/935678348330434570/1035311559993462944/1183974289104441354)
This PR doesn't include deployment script

### CODE REVIEW

**Comment by yorhodes** [comment]
@NOOMA-42 this is looking great, is it ready for review?

**Comment by NOOMA-42** [comment]
> @NOOMA-42 this is looking great, is it ready for review?

it's ready now. 
Got a questions:
I only implement Eth to Polygon. Do I need to implement sending message from polygon to eth? There's something called `sendMessageToRoot` in fxchild contract on Polygon Pos. I'm not sure do should I just implement such method in polygonISM or there should be another hook. I've left some other comments in the code for further discussion.

File: solidity/contracts/hooks/PolygonPosHook.sol [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        return 0;
+    }
+
+    /// @inheritdoc AbstractMessageIdAuthHook
+    function _sendMessageId(
+        bytes calldata metadata,
+        bytes memory payload
+    ) internal override {
+        require(
+            metadata.msgValue(0) == 0,
```
Comment (NOOMA-42) [inline solidity/contracts/hooks/PolygonPosHook.sol:71]: fxchild on don't seems to support msgValue, should I still use default metadata? (with message value)

**Comment by NOOMA-42** [comment]
Ready for review @yorhodes @nambrot @jmrossy

File: solidity/contracts/hooks/PolygonPosHook.sol [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        FxBaseRootTunnel(_cpManager, _fxRoot)
+    {
+        require(
+            Address.isContract(_cpManager),
+            "PolygonPosHook: invalid cpManager contract"
+        );
+        require(
+            Address.isContract(_fxRoot),
+            "PolygonPosHook: invalid fxRoot contract"
+        );
```
Comment (yorhodes) [inline solidity/contracts/hooks/PolygonPosHook.sol:54]: does `FxBaseRootTunnel` not do this?

File: solidity/contracts/hooks/PolygonPosHook.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/contracts/hooks/PolygonPosHook.sol:71]: this is fine
I think we should also `require(msg.value == 0)`

File: solidity/contracts/hooks/PolygonPosHook.sol [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        _sendMessageToChild(payload);
+    }
+
+    // FIX: connect to mailbox, check how to do bidrectional
+    bytes public latestData;
+
+    function _processMessageFromChild(bytes memory data) internal override {
+        latestData = data;
+    }
```
Comment (yorhodes) [inline solidity/contracts/hooks/PolygonPosHook.sol:82]: I dont understand whats going on here

File: solidity/contracts/isms/hook/PolygonPosIsm.sol [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    // ============ Internal function ============
+
+    uint256 public latestStateId;
+    address public latestRootMessageSender;
+    bytes public latestData;
+
+    /* function sendMessageToRoot(bytes memory message) public {
+        _sendMessageToRoot(message);
+    } */
```
Comment (yorhodes) [inline solidity/contracts/isms/hook/PolygonPosIsm.sol:57]: whats going on here?

File: solidity/test/isms/PolygonPosIsm.t.sol [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+            address(l1Mailbox),
+            POLYGON_POS_DOMAIN,
+            TypeCasts.addressToBytes32(address(polygonPosISM)),
+            MAINNET_CHECKPOINT_MANAGER,
+            MAINNET_FX_ROOT
+        );
+
+        polygonPosHook.setFxChildTunnel(address(polygonPosISM));
+
+        vm.makePersistent(address(polygonPosHook));
```
Comment (yorhodes) [inline solidity/test/isms/PolygonPosIsm.t.sol:113]: what does this do?

File: solidity/test/isms/PolygonPosIsm.t.sol [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        emit StateSynced(
+            (stateSender.counter() + 1),
+            MAINNET_FX_CHILD,
+            abi.encode(
+                TypeCasts.addressToBytes32(address(polygonPosHook)),
+                TypeCasts.addressToBytes32(address(polygonPosISM)),
+                encodedHookData
+            )
+        );
+        polygonPosHook.postDispatch(testMetadata, encodedMessage);
```
Comment (yorhodes) [inline solidity/test/isms/PolygonPosIsm.t.sol:176]: nice

File: solidity/contracts/hooks/PolygonPosHook.sol [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    /// @inheritdoc AbstractMessageIdAuthHook
+    function _sendMessageId(
+        bytes calldata metadata,
+        bytes memory payload
+    ) internal override {
+        require(
+            metadata.msgValue(0) == 0,
+            "PolygonPosHook: Fxchild not support msgValue"
```
Comment (yorhodes) [inline solidity/contracts/hooks/PolygonPosHook.sol:72]: nit: 
```suggestion
            "PolygonPosHook: does not support msg.value"
```

**Review by yorhodes — APPROVED** [review]
great work! amazing test coverage
would love to see more contributions from you

File: solidity/contracts/hooks/PolygonPosHook.sol (hunk 1 above)
Comment (NOOMA-42) [reply solidity/contracts/hooks/PolygonPosHook.sol:71]: yes, to make sure no msg.value is passed. I'll add

File: solidity/contracts/hooks/PolygonPosHook.sol (hunk 2 above)
Comment (NOOMA-42) [reply solidity/contracts/hooks/PolygonPosHook.sol:54]: No it's not checked in `FxBaseRootTunnel`. It construct with below 2 interfaces
```
constructor(address _checkpointManager, address _fxRoot) {
        checkpointManager = ICheckpointManager(_checkpointManager);
        fxRoot = IFxStateSender(_fxRoot);
    }
```

File: solidity/contracts/hooks/PolygonPosHook.sol (hunk 3 above)
Comment (NOOMA-42) [reply solidity/contracts/hooks/PolygonPosHook.sol:82]: this comment has been done, i'll remove it

File: solidity/contracts/isms/hook/PolygonPosIsm.sol (hunk 4 above)
Comment (NOOMA-42) [reply solidity/contracts/isms/hook/PolygonPosIsm.sol:57]: ah this is something I'd like to discuss, do i need to implement child to root message? (polygon pos to ethereum) which I don't think this should be implemented in ISM.

File: solidity/test/isms/PolygonPosIsm.t.sol (hunk 5 above)
Comment (NOOMA-42) [reply solidity/test/isms/PolygonPosIsm.t.sol:113]: hook on Ethereum inherents FxRootTunnel and need to set a destination, which is FxChildTunnel on Polygon Pos

**Comment by avious00** [comment]
hey @yorhodes what's remaining for @NOOMA-42 here? 

does he need to resolve the 4 open conversations, resolve conflicts, pass tests, and then merge?

**Pull request merged**